protocol/icmpv6.go

```diff
@@ -0,0 +1,643 @@
+//    |                                                               |
+//    *                                                               *
+//    |                                                               |
+//    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+type MLDQuery struct {
+	ICMPv6Header
+	MaxResponse              uint16
+	Reserved                 uint16
+	MulticastAddress         net.IP
+	Version2                 bool
+	Reserved2                uint8
+	SuppressRouterProcessing bool
+	RobustnessValue          uint8
+	IntervalTime             uint8
+	NumberOfSources          uint16
+	SourceAddresses          []net.IP
+}
+
+func (q *MLDQuery) Len() uint16 {
+	if !q.Version2 {
+		return 28
+	}
+	return 28 + q.NumberOfSources*16
+}
+
+func (q *MLDQuery) MarshalBinary() (data []byte, err error) {
+	data = make([]byte, int(q.Len()))
+	n := 0
+	b, err := q.ICMPv6Header.MarshalBinary()
+	if err != nil {
+		return nil, err
+	}
+	copy(data[n:], b)
+	n += len(b)
+	binary.BigEndian.PutUint16(data[n:], q.MaxResponse)
+	n += 2
+	n += 2
+	copy(data[n:], q.MulticastAddress)
+	n += 16
+	if q.Version2 {
+		sBit := uint8(0x0)
+		if q.SuppressRouterProcessing {
+			sBit = 0x8
+		}
+		data[n] = sBit | q.RobustnessValue&0x7
+		n += 1
+		data[n] = q.IntervalTime
+		n += 1
+		binary.BigEndian.PutUint16(data[n:], q.NumberOfSources)
+		n += 2
+		for _, src := range q.SourceAddresses {
+			copy(data[n:n+16], src.To16())
+			n += 16
+		}
+	}
+	return data, nil
+}
+
+func (q *MLDQuery) UnmarshalBinary(data []byte) error {
+	if len(data) < 4 {
+		return errors.New("The []byte is too short to unmarshal a full ICMPv6Header message.")
+	}
+	err := q.ICMPv6Header.UnmarshalBinary(data)
+	if err != nil {
+		return err
+	}
+	n := q.ICMPv6Header.Len()
+	if len(data) < 28 {
+		return fmt.Errorf("The []byte is too short to unmarshal a full MLDQuery message.")
+	}
+	q.MaxResponse = binary.BigEndian.Uint16(data[n:])
+	n += 2
+	n += 2
+	q.MulticastAddress = data[n : n+16]
+	n += 16
+	q.Version2 = len(data) > 28
+	if q.Version2 {
+		q.SuppressRouterProcessing = data[n]&0x8 != 0
+		q.RobustnessValue = data[n] & 0x7
+		n += 1
+		q.IntervalTime = data[n]
+		n += 1
+		q.NumberOfSources = binary.BigEndian.Uint16(data[n:])
+		n += 2
+		if len(data) < int(q.Len()) {
+			return fmt.Errorf("The []byte is too short to unmarshal a full MLDv2 Query message.")
+		}
+		for j := 0; j < int(q.NumberOfSources); j++ {
+			q.SourceAddresses = append(q.SourceAddresses, data[n:n+16])
+			n += 16
+		}
+	}
+	return nil
+}
+
+func NewMLDQuery(maxResponse uint16, multicastIP net.IP) *MLDQuery {
+	return &MLDQuery{
+		ICMPv6Header: ICMPv6Header{
+			Type: ICMPv6_Type_MLD_Query,
+			Code: 0,
+		},
+		MaxResponse:      maxResponse,
+		MulticastAddress: multicastIP,
+		Version2:         false,
+	}
+}
+
+func NewMLDv2Query(group net.IP, maxResponseTime uint16, queryInterval uint8, sources []net.IP) *MLDQuery {
```

Comment:
Could we use use same variable names and keep same order as `NewMLDQuery`?

Reply:
updated.